lib/pbench/server/api/resources/query_apis/datasets/datasets.py

```diff
@@ -0,0 +1,347 @@
+                authorization=ApiAuthorizationType.DATASET,
+                attributes=ApiAttributes(
+                    "update",
+                    OperationName.UPDATE,
+                    require_stable=True,
+                    require_map=True,
+                ),
+            ),
+            ApiSchema(
+                ApiMethod.DELETE,
+                OperationCode.DELETE,
+                uri_schema=Schema(
+                    Parameter("dataset", ParamType.DATASET, required=True),
+                ),
+                audit_type=AuditType.DATASET,
+                audit_name="delete",
+                authorization=ApiAuthorizationType.DATASET,
+                attributes=ApiAttributes(
+                    "delete",
+                    OperationName.DELETE,
+                    require_stable=True,
+                    require_map=False,
+                ),
+            ),
+            ApiSchema(
+                ApiMethod.GET,
+                OperationCode.READ,
+                uri_schema=Schema(
+                    Parameter("dataset", ParamType.DATASET, required=True)
+                ),
+                authorization=ApiAuthorizationType.DATASET,
+                attributes=ApiAttributes(
+                    "get",
+                    None,
+                    require_stable=False,
+                    require_map=True,
+                ),
+            ),
+        )
+
+    def assemble(self, params: ApiParams, context: ApiContext) -> JSONOBJECT:
+        """Construct an Elasticsearch query to GET/UPDATE/DELETE documents
+
+        This assembly handles three related cases:
+
+        1. To update the access or ownership of a dataset, using the
+           _update_by_query operation to find all indexed documents owned by
+           the specified dataset and update their "authorization" subdocument
+           with a "painless" script.
+        2. To delete a dataset, using the _delete_by_query operation to find
+           and delete all indexed documents owned by the specified dataset.
+        3. An almost-nearly-free diagnostic API to return all indexed documents
+           owned by a specified dataset.
+
+        All three operations are based on the same Elasticsearch query
+        expression, which searches the set of indices containing documents for
+        the specified dataset for documents owned by the dataset resource_id
+        (MD5).
+
+        NOTE: while most index document templates record the owning dataset
+        resource_id as "run.id", the TOC document records it as
+        "run_data_parent", and we need to match against either.
+        """
+
+        dataset = context["dataset"]
+        action = context["attributes"].action
+        get_action = action == "get"
+        context["action"] = action
+        audit_attributes = {}
+        access = None
+        owner = None
+        elastic_options = {"ignore_unavailable": "true"}
+
+        if not get_action:
```

Comment:
True: the code paths underwent some churn before I got it working. Generally I prefer to avoid text comparisons, and I'm not entirely sure why I felt hesitant to just define a constant or an enum here, which might have been more appropriate...